### PR TITLE
Few daemon duplication refactor

### PR DESCRIPTION
**Pull request by vdemeester**
Few small refactoring to limit duplication where I thought I could. For some part this is more a try to get opinion than a definite refactoring. It is separate in separate commit to ease the review and also to remove if the refactoring doesn't make sense.

The refactor happens mainly in `cluster` package and at some part of the `daemon` one.

/cc @aaronlehmann @tonistiigi @AkihiroSuda @yongtang @cpuguy83 

Hopefully this does not break anything 👼

🐸

### CODE REVIEW

File: daemon/cluster/cluster.go [hunk 1]
```diff
@@ -1570,8 +1529,7 @@ func (c *Cluster) CreateNetwork(s apitypes.NetworkCreateRequest) (string, error)
 	return r.Network.ID, nil
 }
 
-// RemoveNetwork removes a cluster network.
-func (c *Cluster) RemoveNetwork(input string) error {
+func (c *Cluster) updateClusterObject(updateFn func(ctx context.Context, state nodeState) error) error {
```
Comment (vdemeester) [inline daemon/cluster/cluster.go]: I think we could do better than that 👼

File: daemon/container.go [hunk 2]
```diff
@@ -54,6 +54,19 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
 	return daemon.containers.Get(containerID), nil
 }
 
+func (daemon *Daemon) GetcontainerCondition(prefixOrName string, conditions ...func(*container.Container) error) (*container.Container, error) {
```
Comment (vdemeester) [inline daemon/container.go]: I don't like the name I used 👼 I should change it 😄

File: daemon/container_operations.go [hunk 3]
```diff
@@ -829,16 +829,24 @@ func (daemon *Daemon) disconnectFromNetwork(container *container.Container, n li
...
+
+func (daemon *Daemon) tryDetachContainerFromClusterNetwork(network libnetwork.Network, container *container.Container) {
+	if daemon.clusterProvider != nil && network.Info().Dynamic() && !container.Managed {
+		if err := daemon.clusterProvider.DetachNetwork(network.Name(), container.ID); err != nil {
+			logrus.Warnf("error detaching from network %s: %v", network.Name(), err)
+			if err := daemon.clusterProvider.DetachNetwork(network.ID(), container.ID); err != nil {
+				logrus.Warnf("error detaching from network %s: %v", network.ID(), err)
 			}
 		}
+		attributes := map[string]string{
```
Comment (vdemeester) [inline daemon/container_operations.go]: This was not on the in the initial "call" (above) and I feel that's a miss.

File: daemon/graphdriver/counter.go [hunk 4]
```diff
@@ -24,29 +24,19 @@ func NewRefCounter(c Checker) *RefCounter {
...
 }
 
 // Decrement decreases the ref count for the given id and returns the current count
 func (c *RefCounter) Decrement(path string) int {
+	return c.operation(path, func(minfo *minfo) {
+		minfo.count--
+	})
+}
+
+func (c *RefCounter) operation(path string, infoOp func(minfo *minfo)) int {
```
Comment (vdemeester) [inline daemon/graphdriver/counter.go]: This one I'm not sure about 😛

**Comment by LK4D4** [comment]
@vdemeester gofmt issue

**Comment by vdemeester** [comment]
@LK4D4 yeah saw that, should be fixed 👼

File: daemon/graphdriver/counter.go [hunk 5]
```diff
@@ -24,29 +24,19 @@ func NewRefCounter(c Checker) *RefCounter {
...
-	return count
+	return c.operation(path, func(minfo *minfo) {
+		minfo.count++
+	})
 }
 
 // Decrement decreases the ref count for the given id and returns the current count
 func (c *RefCounter) Decrement(path string) int {
+	return c.operation(path, func(minfo *minfo) {
+		minfo.count--
```
Comment (AkihiroSuda) [inline daemon/graphdriver/counter.go]: Probably need zero check here
https://github.com/docker/docker/pull/30811

File: daemon/graphdriver/counter.go (hunk 4 above)
Comment (AkihiroSuda) [inline daemon/graphdriver/counter.go]: I wonder it is too much to have general func here.
I think `func (c *RefCounter) incdec(path string, inc bool)) int` is ok

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (allencloud) [inline daemon/cluster/cluster.go]: I am not sure if `updateClusterObject` does not contain the following codes
```
        state := c.currentNodeState()
 	if !state.IsActiveManager() {
 		return c.errNoManager(state)		
 	}
```
will the handler return an error of `this node is not part of swarm`.

Then a docker with no swarm mode will response to cluster update action which may break api.

File: daemon/container.go (hunk 2 above)
Comment (yongtang) [inline daemon/container.go]: The case may be changed, e.g., `s/GetcontainerCondition/GetContainerCondition/`, ?

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (vdemeester) [inline daemon/cluster/cluster.go]: The github diff doesn't really show it properly, but `updateClusterObject` does contain this piece of code 😉

**Comment by aaronlehmann** [comment]
There are still `gofmt` issues.

File: daemon/container.go (hunk 2 above)
Comment (aaronlehmann) [inline daemon/container.go]: Agreed, I see what you're going for with `GetcontainerCondition` but I don't care for it. What about `CheckContainer`?

File: daemon/daemon.go [hunk 6]
```diff
@@ -879,33 +879,30 @@ func (daemon *Daemon) Unmount(container *container.Container) error {
 
 // V4Subnets returns the IPv4 subnets of networks that are managed by Docker.
 func (daemon *Daemon) V4Subnets() []net.IPNet {
```
Comment (aaronlehmann) [inline daemon/daemon.go]: What about refactoring `V4Subnets` and `V6Subnets` into a single function that returns both slices? They are both called in the same place: https://github.com/docker/docker/blob/584e399d9632200e6c8a636339c8aa6347a57fca/daemon/cluster/listen_addr.go#L164-L166

File: daemon/daemon_unix.go [hunk 7]
```diff
@@ -1265,19 +1265,21 @@ func (daemon *Daemon) initCgroupsPath(path string) error {
...
+		return err
 	}
-	if sysinfo.CPURealtimeRuntime && daemon.configStore.CPURealtimeRuntime != 0 {
+	if err := createCPURealTimeFileIfRequired(sysinfo.CPURealtimeRuntime, daemon.configStore.CPURealtimeRuntime, "cpu.rt_runtime_us", path); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createCPURealTimeFileIfRequired(sysinfoPresent bool, configValue int64, file string, path string) error {
```
Comment (aaronlehmann) [inline daemon/daemon_unix.go]: `maybeCreateCPURealTimeFile`?

File: daemon/list.go [hunk 8]
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
 		names:                daemon.nameIndex.GetAll(),
 	}, nil
 }
+func portOp(key string, filter map[nat.Port]bool) func(value string) error {
+	return func(value string) error {
+		if strings.Contains(value, ":") {
+			return fmt.Errorf("filter for '%s' should not contain ':': %v", key, value)
```
Comment (aaronlehmann) [inline daemon/list.go]: Why `%s` for `key` and `%v` for `value` when they are both strings?

File: daemon/list.go [hunk 9]
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
...
+		}
+		//support two formats, original format <portnum>/[<proto>] or <startport-endport>/[<proto>]
+		proto, port := nat.SplitProtoPort(value)
+		start, end, err := nat.ParsePortRange(port)
+		if err != nil {
+			return fmt.Errorf("error while looking up for %s %v: %s", key, value, err)
```
Comment (aaronlehmann) [inline daemon/list.go]: Why `%s` for `key` and `%v` for `value` when they are both strings?

File: daemon/list.go [hunk 10]
```diff
@@ -382,6 +346,27 @@ func (daemon *Daemon) foldFilter(config *types.ContainerListOptions) (*listConte
...
+		}
+		for i := start; i <= end; i++ {
+			p, err := nat.NewPort(proto, strconv.FormatUint(i, 10))
+			if err != nil {
+				return fmt.Errorf("error while looking up for %s %v: %s", key, value, err)
```
Comment (aaronlehmann) [inline daemon/list.go]: Why `%s` for `key` and `%v` for `value` when they are both strings?

File: daemon/cluster/cluster.go (hunk 1 above)
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: I would prefer a name like `lockedManagerAction`.

File: daemon/container.go (hunk 2 above)
Comment (vdemeester) [inline daemon/container.go]: @aaronlehmann for the name of the method, or as a method that would only do the checks ?

File: daemon/daemon.go (hunk 6 above)
Comment (vdemeester) [inline daemon/daemon.go]: ah you're right 👼

File: daemon/list.go (hunk 8 above)
Comment (vdemeester) [inline daemon/list.go]: Didn't change the code inside (just extracted it), but you are right, we could use `%s` for both, I'll update :wink:

File: daemon/container.go (hunk 2 above)
Comment (aaronlehmann) [inline daemon/container.go]: The name of the method.

**Comment by vdemeester** [comment]
Updated 👼

File: daemon/daemon.go [hunk 11]
```diff
@@ -879,40 +879,28 @@ func (daemon *Daemon) Unmount(container *container.Container) error {
...
 
 	for _, managedNetwork := range managedNetworks {
-		v4Infos, _ := managedNetwork.Info().IpamInfo()
-		for _, v4Info := range v4Infos {
-			if v4Info.IPAMData.Pool != nil {
-				subnets = append(subnets, *v4Info.IPAMData.Pool)
+		v4infos, v6infos := managedNetwork.Info().IpamInfo()
+		for _, info := range v4infos {
+			if info.IPAMData.Pool != nil {
+				v4Subnets = append(v6Subnets, *info.IPAMData.Pool)
```
Comment (aaronlehmann) [inline daemon/daemon.go]: Wrong argument to `append`.

Comment (vdemeester) [inline daemon/daemon.go]: ah damn :sweat_smile:

File: daemon/cluster/networks.go [hunk 12]
```diff
@@ -211,51 +208,38 @@ func (c *Cluster) DetachNetwork(target string, containerID string) error {
...
-	r, err := state.controlClient.CreateNetwork(ctx, &swarmapi.CreateNetworkRequest{Spec: &networkSpec})
-	if err != nil {
+	var resp *swarmapi.CreateNetworkResponse
+	if err := c.lockedManagerAction(func(ctx context.Context, state nodeState) error {
+		networkSpec := convert.BasicNetworkCreateToGRPC(s)
+		r, err := state.controlClient.CreateNetwork(ctx, &swarmapi.CreateNetworkRequest{Spec: &networkSpec})
+		if err != nil {
+			return err
+		}
+		resp = r
```
Comment (AkihiroSuda) [inline daemon/cluster/networks.go]: nit: you don't need `r`

File: daemon/cluster/services.go [hunk 13]
```diff
@@ -59,207 +59,181 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
...
-			resp.Warnings = append(resp.Warnings, fmt.Sprintf("unable to pin image %s to digest: %s", newCtnr.Image, err.Error()))
-		} else if newCtnr.Image != digestImage {
-			logrus.Debugf("pinning image %s by digest: %s", newCtnr.Image, digestImage)
-			newCtnr.Image = digestImage
-		} else {
-			logrus.Debugf("updating service using supplied digest reference %s", newCtnr.Image)
+		resp := &apitypes.ServiceUpdateResponse{}
+
+		// pin image by digest
+		if os.Getenv("DOCKER_SERVICE_PREFER_OFFLINE_IMAGE") != "1" {
```
Comment (AkihiroSuda) [inline daemon/cluster/services.go]: similar to L110-121, can we deduplicate? (in future PR)

File: daemon/container.go [hunk 14]
```diff
@@ -54,6 +54,16 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
 	return daemon.containers.Get(containerID), nil
 }
 
+// CheckContainer make sure the specified container validates the specified conditions
+func (daemon *Daemon) CheckContainer(container *container.Container, conditions ...func(*container.Container) error) error {
```
Comment (tonistiigi) [inline daemon/container.go]: we could probably leave this method private

File: daemon/container_operations_unix.go [hunk 15]
```diff
@@ -58,32 +58,34 @@ func (daemon *Daemon) setupLinkedContainers(container *container.Container) ([]s
...
 	if err != nil {
 		return nil, err
 	}
+	return container, daemon.CheckContainer(container, containerIsRunning, containerIsNotRestarting)
+}
+
+func containerIsRunning(c *container.Container) error {
 	if !c.IsRunning() {
-		return nil, fmt.Errorf("cannot join PID of a non running container: %s", containerID)
+		return fmt.Errorf("cannot join IPC of a non running container: %s", c.ID)
```
Comment (tonistiigi) [inline daemon/container_operations_unix.go]: This message would be also shown on pidns error

File: daemon/container.go (hunk 14 above)
Comment (vdemeester) [inline daemon/container.go]: right 👼

File: daemon/container_operations_unix.go (hunk 15 above)
Comment (vdemeester) [inline daemon/container_operations_unix.go]: ah you are right, I need to handle errors differently 👼

**Comment by vdemeester** [comment]
@tonistiigi updated 👼

**Comment by tonistiigi** [comment]
LGTM

**Review by AkihiroSuda — APPROVED** [review]
LGTM, my above comments are nitpicky :smile:


I think this PR can be merged without squashing, but deferring to somebody else for confirmation :sweat_smile:

File: daemon/container.go [hunk 16]
```diff
@@ -54,8 +54,8 @@ func (daemon *Daemon) GetContainer(prefixOrName string) (*container.Container, e
 	return daemon.containers.Get(containerID), nil
 }
 
-// CheckContainer make sure the specified container validates the specified conditions
-func (daemon *Daemon) CheckContainer(container *container.Container, conditions ...func(*container.Container) error) error {
+// checkContainer make sure the specified container validates the specified conditions
+func (daemon *Daemon) checkContainer(container *container.Container, conditions ...func(*container.Container) error) error {
```
Comment (thaJeztah) [inline daemon/container.go]: These checkContainer changes are in the wrong commit; should be in f5c490ac633c25ef5f83e99e524a954d50801fb5

File: daemon/container_operations_unix.go [hunk 17]
```diff
@@ -60,23 +60,23 @@ func (daemon *Daemon) getIpcContainer(container *container.Container) (*containe
 	containerID := container.HostConfig.IpcMode.Container()
 	container, err := daemon.GetContainer(containerID)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "cannot join IPC of a non running container: %s", container.ID)
 	}
-	return container, daemon.CheckContainer(container, containerIsRunning, containerIsNotRestarting)
+	return container, daemon.checkContainer(container, containerIsRunning, containerIsNotRestarting)
```
Comment (thaJeztah) [inline daemon/container_operations_unix.go]: Same for these :smile:

**Comment by vdemeester** [comment]
@thaJeztah updated 👼

File: daemon/container_operations_unix.go [hunk 18]
```diff
@@ -69,14 +69,14 @@ func (daemon *Daemon) getPidContainer(container *container.Container) (*containe
...
-		return nil, err
+		return nil, errors.Wrapf(err, "cannot join PID of a non running container: %s", container.ID)
 	}
 	return container, daemon.checkContainer(container, containerIsRunning, containerIsNotRestarting)
 }
 
 func containerIsRunning(c *container.Container) error {
 	if !c.IsRunning() {
-		return fmt.Errorf("cannot join IPC of a non running container: %s", c.ID)
+		return errors.Errorf("container %s is not running", c.ID)
```
Comment (thaJeztah) [inline daemon/container_operations_unix.go]: Actually, looks like these changes also belong in 25caa5e9194db2cee946e9ef98557dca9976f6d3

https://github.com/docker/docker/pull/30843/commits/25caa5e9194db2cee946e9ef98557dca9976f6d3#r102823906

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**